engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/internal/PairwiseFloatRingBuffer.java

```diff
@@ -0,0 +1,381 @@
+            final float rightVal = storageChunk.get(right);
+            final float parentVal = storageChunk.get(parent);
+
+            final float computeVal = pairwiseFunction.apply(leftVal, rightVal);
+            if (parentVal != computeVal) {
+                storageChunk.set(parent, computeVal);
+                // mark the parent dirty (if not the last)
+                if (parent > 1) {
+                    dirtyIndices.add(parent);
+                }
+            }
+            // how far should we advance
+            final int nextIndex = dirtyIndex + 1;
+            if (nextIndex < dirtyIndices.size() && dirtyIndices.get(nextIndex) == right) {
+                dirtyIndex += 2;
+            } else {
+                dirtyIndex++;
+            }
+        }
+        allDirty = false;
+        dirtyIndices.clear();
+
+        // final value is in index 1
+        return storageChunk.get(1);
+    }
+
+    private void grow(int increase) {
+        int oldCapacity = capacity;
+        int oldChunkSize = chunkSize;
+
+        int size = size();
+
+        final int minLength = size + increase;
+
+        // double the current capacity until there is sufficient space for the increase
+        while (capacity <= minLength) {
+            capacity *= 2;
+            chunkSize = capacity * 2;
+        }
+
+        // transfer to the new chunk
+        WritableFloatChunk<Values> oldChunk = storageChunk;
+        storageChunk = WritableFloatChunk.makeWritableChunk(chunkSize);
+
+        // fill the pairwise tree (0 to capacity) with empty value
+        storageChunk.fillWithValue(0, capacity, emptyVal);
+
+        // move the data to the new chunk, note that we store the ring data in the second half of the array
+
+        if (tail >= head) {
+            storageChunk.copyFromTypedChunk(oldChunk, head, capacity, size);
+        } else {
+            final int firstCopyLen = oldChunkSize - head;
+            storageChunk.copyFromTypedChunk(oldChunk, head, capacity, firstCopyLen);
+            storageChunk.copyFromTypedChunk(oldChunk, oldCapacity, capacity + firstCopyLen , size - firstCopyLen);
+        }
+        tail = capacity + size;
+
+        // fill the unused storage with the empty value
+        storageChunk.fillWithValue(tail, chunkSize - tail, emptyVal);
+
+        // free the old data chunk
+        oldChunk.close();
+        head = capacity;
+
+        // TODO: investigate moving precomputed results also.  Since we are re-ordering the data values, would be
+        // tricky to maintain order but a recursive function could probably do it efficiently.  For now, make life easy
+        // by setting all input dirty so the tree is recomputed on next `evaluate()`
+        this.dirtyIndices.clear();
+        allDirty = true;
+    }
+
+    private void grow() {
+        grow(1);
+    }
+
+    public void push(float val) {
+        if (isFull()) {
+            grow();
+        }
+        // add the new data
```

Comment:
Replace the guts of `push` and `pop` with a call to `pushUnsafe` or `popUnsafe` as appropriate.